[llvm] update to v21.1.8

yurybura opened the pull request:
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] SHA512s are updated for each updated download.
- [ ] The "supports" clause reflects platforms that may be fixed by this new version.
- [ ] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

Fixes #44388
Fixes #44249
Fixes #43595

Fixes #33576

yurybura commented:
> For a casual lurker - what's the justification for skipping 19.x and 20.x?

No one has updated these versions yet. Due to lack of free time, I'll be focusing on updating to the latest version.

kykrueger commented on versions/l-/llvm.json:
```diff
@@ -1,5 +1,10 @@
 {
   "versions": [
+    {
+      "git-tree": "fd53a0affe21c7dd537dee100c2c8547cbd034a1",
```
```suggestion
      "git-tree": "0dae76a75b246b7cd35bd8e7167a1ad3647f959e",
```

lriggs commented:
@yurybura Anything I can do to help? Very interested in this update.

NekkoDroid commented:
I also want to point out to the request in #37594 since it might make sense to bundle the change together.

yurybura commented on ports/halide/portfile.cmake:
```diff
@@ -1,64 +1,35 @@
-        -DCMAKE_INSTALL_LIBDIR=bin
-        "-DCMAKE_INSTALL_DATADIR=share/${PORT}"
-        "-DHalide_INSTALL_CMAKEDIR=share/${PORT}"
+        -DWITH_AUTOSCHEDULERS=OFF
```
Okay, got this. I've disabled it temporarily.

yurybura commented:
> I also want to point out to the request in #37594 since it might make sense to bundle the change together.

I've seen this issue. It's very difficult to test all the features because the build is so slow. I think it's better to split LLVM and use separate builds for individual subprojects. I think I can do this in the next pull request. Please wait)

alexreinking commented on ports/halide/vcpkg.json:
```diff
@@ -55,151 +25,5 @@
-        }
-      ]
-    },
-    "target-all": {
-      "description": "Include all targets",
-      "dependencies": [
-        {
-          "name": "halide",
-          "default-features": false,
-          "features": [
-            "target-aarch64",
-            "target-amdgpu",
-            "target-arm",
-            "target-d3d12compute",
-            "target-hexagon",
-            "target-metal",
-            "target-nvptx",
-            "target-opencl",
-            "target-opengl-compute",
-            "target-powerpc",
-            "target-riscv",
-            "target-webassembly",
-            "target-x86"
-          ]
-        }
-      ]
-    },
```
Ah, I'm sorry -- I forgot that we changed how this works in halide/Halide#8379. The way you have it implemented now is fine as long as Halide will rebuild if the LLVM features change.

alexreinking replied:
I still think it could be useful to have the LLVM features exposed on Halide so you can ask for particular support in Halide explicitly. It's simpler to say your project depends on `halide[target-riscv]` if that's what you're working on than to list it on LLVM, assuming Halide will pick it up.

yurybura committed:
[llvm] update to v21.1.8
- remove enable-terminfo, pstl
- add clang-enable-cir, clang-enable-static-analyzer
- support ARM on Windows
- support cross-compiling

yurybura committed:
[symengine] update to v0.14.0

yurybura committed:
[halide] update to v21.0.0

yurybura commented:
> Using a local overlay worked well for me, thanks. Is there any reason to not use the overlays in our build system? ie is the underlying llvm package stable, even without this PR getting merged?

I think the LLVM update is quite stable. But to complete this PR, I need to update other dependent libraries and test all the features. On my computer, re-building LLVM can take up to half a day, which slows things down considerably. I plan to return to this update as soon as I have the opportunity. But the update will be more complex - I want to split LLVM into subprojects and build them independently. As a result, you won't have to completely rebuild LLVM with all the projects each time, but you'll be able to build them independently.

yurybura committed:
[libmem] support new LLVM

xavier2k6 commented:
Latest `libmem` (attow) is `5.1.4`
https://github.com/rdbo/libmem/releases/tag/5.1.4

r-tmp commented:
Don't mean to pester, but LLVM v22 is around the corner and vcpkg is stuck in v18, what's stopping this from being merged?

yurybura commented:
> Don't mean to pester, but LLVM v22 is around the corner and vcpkg is stuck in v18, what's stopping this from being merged?

My apologies, but the main reason is time... To complete this pull request, I need to test all LLVM functions (all combinations) and all dependent projects.

My laptop isn't fast enough for this to take a reasonable amount of time. I think it's simpler and more efficient to split LLVM into subprojects and build them independently. I'll get back to this task as soon as I can.
